File: lab4/linked_list_circular.cpp
#include <iostream>

using namespace std;

class Node {
public:
    int data;
    Node* next;

    explicit Node(int value) : data(value), next(nullptr) {}
};

class CQueue {
private:
    Node* front;
    Node* rear;

public:
    CQueue() : front(nullptr), rear(nullptr) {}
    ~CQueue() {
        while (front != nullptr) {
            deleteFront();
        }
    }

    void insertion();
    void deletion();
    void display();
    void showTailNext();
    void deleteFront();
};

int main() {
    CQueue cqobj;
    int ch;

    do {
        cout << "\n\n\tMain Menu";
        cout << "\n##########################";
        cout << "\n1. Insert\n2. Delete\n3. Display\n4. Show tail->next\n5. Exit\n\nEnter Your Choice: ";
        cin >> ch;

        switch (ch) {
            case 1:
                cqobj.insertion();
                cqobj.display();
                break;
            case 2:
                cqobj.deletion();
                break;
            case 3:
                cqobj.display();
                break;
            case 4:
                cqobj.showTailNext();
                break;
            case 5:
                break;
            default:
                cout << "\n\nWrong Choice!!! Try Again.";
        }
    } while (ch != 5);

    return 0;
}

void CQueue::insertion() {
    int value;
    cout << "\nEnter the Element: ";
    cin >> value;

    Node* n = new Node(value);

    if (front == nullptr) {
        front = rear = n;
        rear->next = front;
    } else {
        rear->next = n;
        rear = n;
        rear->next = front;
    }
}

void CQueue::deleteFront() {
    if (front == nullptr) {
        return;
    }

    if (front == rear) {
        delete front;
        front = rear = nullptr;
    } else {
        Node* temp = front;
        front = front->next;
        rear->next = front;
        delete temp;
    }
}

void CQueue::deletion() {
    if (front == nullptr) {
        cout << "\nCircular Queue Empty!!!";
        return;
    }

    int x = front->data;
    deleteFront();
    cout << "\nElement " << x << " is Deleted";
    display();
}

void CQueue::display() {
    if (front == nullptr) {
        cout << "\n\nCircular Queue Empty!!!";
        return;
    }

    cout << "\n\nCircular Queue Elements are:\n\n";
    Node* temp = front;

    do {
        cout << temp->data << "  ";
        temp = temp->next;
    } while (temp != front);
}

void CQueue::showTailNext() {
    if (rear == nullptr) {
        cout << "Queue is empty" << endl;
        return;
    }

    cout << "Tail->Next->data: " << rear->next->data << endl;
}

File: lab4/linked_list_circular.py
from dataclasses import dataclass


@dataclass
class Node:
    key: int | None = None
    next: "Node | None" = None


class LinkedListCircular:
    """Circular linked list buffer implementation."""

    def __init__(self):
        self.head = None
        self.tail = None

    def len(self) -> int:
        length = 0
        if self.head is not None:
            new_node = self.head
            length += 1
            while new_node.next is not self.head:
                new_node = new_node.next
                length += 1
        return length

    def add(self, key):
        new_node = Node(key)

        if self.head is None:
            self.head = self.tail = new_node
            self.tail.next = self.head
        else:
            self.tail.next = new_node
            self.tail = new_node
            self.tail.next = self.head

    def show(self):
        if self.head is None:
            print("Your queue is empty")
            return

        print("Your queue now is:")
        new_node = self.head
        while new_node is not self.tail:
            print(new_node.key)
            new_node = new_node.next
        print(self.tail.key)

    def show_first(self):
        if self.head is not None:
            print("The first element of queue is:", self.head.key)
        else:
            print("The queue is empty")

    def remove(self):
        if self.head is None:
            print("The queue is empty")
            return

        if self.head == self.tail:
            self.head = self.tail = None
        else:
            self.head = self.head.next
            self.tail.next = self.head


def main():
    ll = LinkedListCircular()

    for i in range(0, 11):
        ll.add(i)

    ll.show()
    print("The size of your queue now is:", ll.len())
    ll.show_first()

    print("Removing the first element of queue:")
    ll.remove()

    ll.show()
    ll.show_first()


if __name__ == "__main__":
    main()
